### ISSUE TITLE
Reproduce  results

### ISSUE THREAD

**Issue by zhibotian** [LiheYoung/UniMatch#5]
### Hi, thanks for your excellent work,  I have some problems in reproduce the result:

- when I reproduced Unimatch on **Voc(1464)**, I could not get the highest result of **81.2**, and I also couldn't get the result in the paper when I used training size as **513*513** for **1/16(662)** and so on, is that because the environment or other reasons? 
- By the way,  I see the code has **OHEM** loss`criterion_l =  ProbOhemCrossEntropy2d(**cfg['criterion']['kwargs']).cuda(local_rank)`, but it doesn't mention it on the paper, could you please tell me the details to how to use **OHEM** loss? 
Thank you very much!

**Comment by LiheYoung** [comment]
- Previously, we have re-run our Pascal experiments three times under each data partition. Below are the results and corresponding training logs. Not that, all results are obtained with ResNet-101 and training size 321. The reported results in our paper are listed in the last row for clear comparison. I think it is easy to reproduce them. Any PyTorch version >= 1.8 should be okay.

| Pascal | 92 | 183 | 366 | 732 | 1464 | 1/16 | 1/8 | 1/4 |
| :-----| :-------| :-------| :-------| :-------| :------| :-------| :------| :------|
| Run 1  | [75.46](https://drive.google.com/file/d/1A_KtVXnXNM1m-b8TRD8199kMI4rpDuzr/view?usp=sharing) | [76.75](https://drive.google.com/file/d/1uRoJ2Cwlr88fGipTctS4N8pkDwem4I04/view?usp=sharing) | [78.51](https://drive.google.com/file/d/1fJkmx6-8k23iQlkNZPu4yVLJBsW4Tn3a/view?usp=sharing) | [79.98](https://drive.google.com/file/d/1WzOXhkwdzPls2wNFVHbaKzaLp1RZgdim/view?usp=sharing) | [80.78](https://drive.google.com/file/d/1qLHxsFrTCtQOV4w_LL5CSPKco6rNta-P/view?usp=sharing) | [76.91](https://drive.google.com/file/d/14GnPaZRPKNpIFyv_2kKOLuwin_MRrlKZ/view?usp=sharing) | [77.46](https://drive.google.com/file/d/12b-6gQt6SIh_gVeSNVHnZaktCEK72CCp/view?usp=sharing) | [77.54](https://drive.google.com/file/d/1rpeim1iCV--2_ngpDO_YG-0tubIAyh6Z/view?usp=sharing) |
| Run 2 | [74.78](https://drive.google.com/file/d/1C5gDZXJQoYqwsYl0VeG5Ov7Tw4fP9d8h/view?usp=sharing) | [78.17](https://drive.google.com/file/d/1SYLUr0_P1lo8wXYeunVLCDh9FO6HyzLx/view?usp=sharing) | [78.69](https://drive.google.com/file/d/1QtkVsc1Yk5tMOIIg-jJ0E4VnxmmoNCfQ/view?usp=sharing) | [80.21](https://drive.google.com/file/d/1if4aVAAdOsnCA-LB-meZA1JVGlgJqKmj/view?usp=sharing) | [80.71](https://drive.google.com/file/d/1NxAfy80KUQSFXQz_6JBoBd53aBlySrE5/view?usp=sharing) | [76.98](https://drive.google.com/file/d/1DM55pHCt0XcRmfLuqaxAAntFc2vBxS1g/view?usp=sharing) | [77.74](https://drive.google.com/file/d/1Fk9S43dMdhrr3yVcKjKqcE9mVQ9RzKTM/view?usp=sharing) | [77.21](https://drive.google.com/file/d/1p2z-MezFjSrmB46Y-SN2qzecIxZIKscU/view?usp=sharing) |
| Run 3 | [74.74](https://drive.google.com/file/d/11faLIcZkDWWHbSwt0w2in3e-IhiiCyg6/view?usp=sharing) | [77.69](https://drive.google.com/file/d/12tuxwrKolVwZ0cSkxvxpB1-niqcTgV6U/view?usp=sharing) | [78.42](https://drive.google.com/file/d/10bHmTf8u_lEmciRhbo-jDNa7Ordb-4Ap/view?usp=sharing) | [79.48](https://drive.google.com/file/d/1ZfGTVn3c0JT6QV5nSEuULFiGWDKgnvC4/view?usp=sharing) | [81.01](https://drive.google.com/file/d/1V15EODnIpSYeE1R2B0fSB5FWE1HbYMQW/view?usp=sharing) | [77.15](https://drive.google.com/file/d/1DYD38i2X0i5HrULGDpNkx4FoB1N9tK84/view?usp=sharing) | [77.05](https://drive.google.com/file/d/1juAWhiMqVz5L-AaAAmKvy5g2Ownx2Ibt/view?usp=sharing) | [77.44](https://drive.google.com/file/d/1g4s-mzxGFoe6jOfXo5yQTq-l3UwBzeQv/view?usp=sharing) |
| Mean (std) | 74.99 (0.40) | 77.54 (0.72) | 78.54 (0.14) | 79.89 (0.37) | 80.83 (0.16) | 77.01 (0.12) | 77.42 (0.35) | 77.40 (0.17) |
||
| Paper | 75.2 | 77.2 | 78.8 | 79.9 | 81.2 | 76.5 | 77.0 | 77.2 |

- Indeed, we have clarified the online hard example mining loss (OHEM) in the Cityscapes paragraph of Section 4.2 in our main paper. To use it, you can follow the Cityscapes config file.

**Comment by LiheYoung** [comment]
I will close this issue. Feel free to re-open it if you have any further questions.

**Comment by zhibotian** [comment]
Hi, as your method in the paper, you also do experiments to use  **Dual-Stream Feature-Level Perturbations**, can you tell me the details about it, using two different weak perturbations feeding to the network to get different Feature-Level perturbations results or just use one weak perturbations but return two after the encoder g. Thanks.

**Comment by LiheYoung** [comment]
It is the latter one as you described. We only use a single weak view, but perform **twice** random feature-level perturbations (Dropout) on it.

Briefly, the code should be changed from:
https://github.com/LiheYoung/UniMatch/blob/7292bace1bda8f64ca05ae42ca8c0eef045aa392/model/semseg/deeplabv3plus.py#L45-L51 to:
```python
if need_fp:
    outs = self._decode(torch.cat((c1, nn.Dropout2d(0.5)(c1), nn.Dropout2d(0.5)(c1))),
                        torch.cat((c4, nn.Dropout2d(0.5)(c4), nn.Dropout2d(0.5)(c4))))
    outs = F.interpolate(outs, size=(h, w), mode="bilinear", align_corners=True)
    out, out_fp1, out_fp2 = outs.chunk(3)

    return out, out_fp1, out_fp2
```

Besides, the original loss function ``loss_u_w_fp`` should also be averaged on the dual views of ``out_fp1`` and ``out_fp2``.

**Comment by zhibotian** [comment]
Okay, get it, thank you very much for your patient reply.

**Comment by zhibotian** [comment]
Hi, Sorry to bother you again, Can you tell me the purpose of this code `ignore_mask[mask == 254] = 255` and the role of the variable `ignore_mask` in the code in `semi.py`, thanks a lot.

**Comment by LiheYoung** [comment]
The ``ignore_mask`` is to mark:
(1) ignored region provided in Pascal GT mask,
(2) padded region during pre-processing. 

The ``mask == 254`` is indeed our padded region (see [here](https://github.com/LiheYoung/UniMatch/blob/a856ccfccd8d637f522d7ee395a41226eb50c4c3/dataset/semi.py#L43-L44)). It does not use value ``255`` for unlabeled images in the first place because we want to tell it from the officially provided ``255`` values in GT masks, which we should not leverage on unlabeled images.

**Comment by zhibotian** [comment]
![1676951312958](https://user-images.githubusercontent.com/73222501/220242843-0c9157c9-b076-44d5-9aab-b23b687b369a.jpg)
Hello, I noticed that the findings of `Resnet101` spilt `1/6, 1/8, 1/4`, for U2PL in your work deviate from the original paper; have you recreated this result? (The first image is the outcome of your paper.)![1676951714776](https://user-images.githubusercontent.com/73222501/220243709-7b9f2bca-7efd-4313-875c-468105b035d4.png)

**Comment by LiheYoung** [comment]
Please refer to https://github.com/Haochen-Wang409/U2PL/issues/3 for detail.

**Comment by zhibotian** [comment]
Soga!  Thanks.
